java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/types/MacAddress.java

```diff
@@ -94,6 +97,21 @@ public static MacAddress of(@Nonnull final String macString) throws IllegalArgum
         return MacAddress.of(raw);
     }
 
+    /**
+     * Creates a {@link MacAddress} from a {@link DatapathId}. This factory
+     * method assumes that the {@link MacAddress} is composed of the last six
+     * bytes of the supplied {@link DatapathId}.
+     * @param dpid the {@link DatapathId} to create the {@link MacAddress} from
+     * @return a {@link MacAddress} derived from the supplied {@link DatapathId}
+     */
+    public static MacAddress of(@Nonnull DatapathId dpid) {
+        Preconditions.checkNotNull(dpid, "dpid must not be null");
+
+        byte[] dpidBytes = dpid.getBytes();
```

Comment:
Why convert via byte array? Both those data structures internally store Dpid/Mac as a long, so you could just retrieve the long and avoid the byte array allocations / copyOf.

Also, should check here that the first two bytes of the DatapathIds are 00:00. Only then can it be represented as a MacAddress.